### PR TITLE
Improve conflict handling gui

### LINKED ISSUE #1195: Unusable Errors/warnings and Conflict messages
Error exists with Final Version 2.5.1. The Client shows a yellow waring message, "Conflicts exist, click here for details" or in German "Es existieren ungelöste Konflikte, für Details klicken".

However it is not possible to click somewhere on the yellow bar (at least nothing happens). The log is also not helpful to debug the messages.

I am using OSX Version 10.13.6.

Hope this could be fixed, since it is annoying to not find out which files are conflicted.

### Expected behaviour
The Button should show a list of files/paths on the machine which are conflicted, this way people can take a look which version they would keep.

### Actual behaviour
Nothing happens if the yellow bar is clicked.

### Steps to reproduce
see above.

### Client configuration
Client version: 2.5.1

Operating system: OSX High Sierra 10.13.6

OS language: German

Qt version used by client package (Linux only, see also Settings dialog): No Linux

Client package (From Nextcloud or distro) (Linux only): No Linux

Installation path of client:
No Idea, since I used the dmg or installer and OSX is not quiet transparent for me in this point, here I miss linux ;(

### PR DESCRIPTION
**Pull request by er-vin**
Fix #1195
Also adopts commits from owncloud/client#6541

As for the conflict dialog itself it looks like this on a Plasma desktop:
Screenshot_20201008_163644

On system for which there's no icon theme by default (typically Windows and Mac) I'm using a fallback icon provided in Qt resources. Here is how it looks:
Screenshot_20201008_155105

I guess shipping a full icon theme (or at least the mimetype icons from said theme...) is an option but honestly that's for later. :-)

Note that clicking the icons (they are buttons which raise on over) also open the corresponding file for preview (just like the more discoverable links).

### CODE REVIEW

**Review by jancborchardt — APPROVED** [review]
Looking good for improving conflict resolution! :) For sure some things we can fix with the spacing, but let's do that separately to get this in quickly.

**Review by jancborchardt — APPROVED** [review]
Looks super nice! Great work @er-vin – will hopefully make it a more relaxing experience for anyone who is stressed cause there’s conflicts. :)

File: src/common/syncjournaldb.cpp [hunk 1]
```diff
@@ -2031,6 +2031,22 @@ QByteArrayList SyncJournalDb::conflictRecordPaths()
...
+    auto conflict = conflictRecord(conflictName);
+    QByteArray result;
+    if (conflict.isValid()) {
+        getFileRecordsByFileId(conflict.baseFileId, [&result](const SyncJournalFileRecord &record) {
+            if (!record._path.isEmpty())
+                result = record._path;
+        });
+    }
+
+    if (result.isEmpty())
```
Comment (camilasan) [inline src/common/syncjournaldb.cpp:2045]: I would use {} even if it is just one line.

**Review by camilasan — CHANGES_REQUESTED** [review]

File: src/common/syncjournaldb.cpp (hunk 1 above)
Comment (er-vin) [reply src/common/syncjournaldb.cpp:2045]: Done

File: src/common/syncjournaldb.h [hunk 2]
```diff
@@ -225,6 +225,13 @@ class OCSYNC_EXPORT SyncJournalDb : public QObject
     /// Return all paths of files with a conflict tag in the name and records in the db
     QByteArrayList conflictRecordPaths();
 
+    /** Find the base name for a conflict file name, using journal or name pattern
+     *
+     * The path must by sync-folder relative.
```
[the remark this answers is not in the document]
Comment (er-vin) [reply src/common/syncjournaldb.h:230]: Done

File: src/gui/socketapi.cpp [hunk 3]
```diff
@@ -689,6 +689,68 @@ void SocketApi::copyUrlToClipboard(const QString &link)
...
+    QString defaultDirAndName = fileData.folderRelativePath;
+
+    // If it's a conflict, we want to save it under the base name by default
+    if (Utility::isConflictFile(defaultDirAndName)) {
+        defaultDirAndName = fileData.folder->journalDb()->conflictFileBaseName(fileData.folderRelativePath.toUtf8());
+    }
+
+    // If the parent doesn't accept new files, go to the root of the sync folder
+    QFileInfo fileInfo(localFile);
+    auto parentRecord = parentDir.journalRecord();
```
Comment (camilasan) [inline src/gui/socketapi.cpp:741]: could this be const?

File: src/gui/socketapi.cpp [hunk 4]
```diff
@@ -689,6 +689,68 @@ void SocketApi::copyUrlToClipboard(const QString &link)
...
+    if ((fileInfo.isFile() && !parentRecord._remotePerm.hasPermission(RemotePermissions::CanAddFile))
+        || (fileInfo.isDir() && !parentRecord._remotePerm.hasPermission(RemotePermissions::CanAddSubDirectories))) {
+        defaultDirAndName = QFileInfo(defaultDirAndName).fileName();
+    }
+
+    // Add back the folder path
+    defaultDirAndName = QDir(fileData.folder->path()).filePath(defaultDirAndName);
+
+    auto target = QFileDialog::getSaveFileName(
```
Comment (camilasan) [inline src/gui/socketapi.cpp:750]: here too?

File: src/gui/socketapi.cpp [hunk 5]
```diff
@@ -814,6 +882,50 @@ void SocketApi::command_GET_MENU_ITEMS(const QString &argument, OCC::SocketListe
         }
 
         sendSharingContextMenuOptions(fileData, listener, !isE2eEncryptedPath);
+
+        // Conflict files get conflict resolution actions
+        bool isConflict = Utility::isConflictFile(fileData.folderRelativePath);
+        if (isConflict || !isOnTheServer) {
+            // Check whether this new file is in a read-only directory
+            QFileInfo fileInfo(fileData.localPath);
+            auto parentDir = fileData.parentFolder();
```
Comment (camilasan) [inline src/gui/socketapi.cpp:900]: const until line 896?

File: src/gui/socketapi.cpp (hunk 3 above)
Comment (er-vin) [reply src/gui/socketapi.cpp]: done

File: src/gui/socketapi.cpp (hunk 5 above)
Comment (er-vin) [reply src/gui/socketapi.cpp]: done

File: src/gui/socketapi.cpp (hunk 4 above)
Comment (er-vin) [reply src/gui/socketapi.cpp]: done

**Review by camilasan — APPROVED** [review]

**Pull request merged**